core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -853,8 +857,11 @@ private[spark] class TaskSchedulerImpl(
     // (taskId, stageId, stageAttemptId, accumUpdates)
     val accumUpdatesWithTaskIds: Array[(Long, Int, Int, Seq[AccumulableInfo])] = {
       accumUpdates.flatMap { case (id, updates) =>
-        val accInfos = updates.map(acc => acc.toInfo(Some(acc.value), None))
         Option(taskIdToTaskSetManager.get(id)).map { taskSetMgr =>
+          val (accInfos, taskProgressRate) = getTaskAccumulableInfosAndProgressRate(updates)
```

Comment:
I'm a bit worried about the scheduler's throughput if our concerns on the accumulators' traverse efficiency matter. I still think we could only traverse inside the speculation thread to decouple with the scheduling thread. If we move this stuff to the speculation thread, we can also avoid unnecessary traverses since it's only necessary when `checkSpeculatableTasks` requires, while with the current implementation it traverses for each heartbeat update and successful task completion.


If we want to move it to the speculation thread, the implementation could be also a bit simpler. At `TaskSchedulerImpl.executorHeartbeatReceived()`, we should only set `_accumulables`. And we don't need to set `_accumulables` by us, which is already covered by `DAGScheudler.updateAccumulators()`. Then, we'd only need to focus on the calculation/traverses at `InefficientTaskCalculator`. It might be a bit slow for the first-time traverses but we can cache the records/runtime for the finished tasks or progress rate for the running tasks. And even if it's slow, I think it's still better compared to slow the scheduling threads. 

@weixiuli @mridulm WDYT?

Reply:
The cost for `getTaskAccumulableInfosAndProgressRate` should be similar to what we currently have already, right ?
This is replacing an existing `updates.map` - though there is an additional `if` check included - are you concerned about that ?

Reply:
Having said that, I was thinking along lines of the following:

1. For successful tasks, as part of TSM.`handleSuccessfulTask`, update the metrics required for efficiency computation.
   1. That is, in `handleSuccessfulTask`, we maintain state required for efficiency computation - without needing to recompute it for all tasks in `maybeRecompute` -> just do it for running tasks.
   2. This is similar to what we do with medianHeap.
2. With this in place, I was not sure if we want to maintain the state in `TaskInfo` anymore.
   1. We could simply maintain a concurrent map for running tasks -> efficiency details in `InefficientTaskCalculator`
   2. In `executorHeartbeatReceived`, we could simply delegate to TSM to update its calculator with the state given task id.
   
Thoughts ?

Reply:
Sounds good. I'll update it. Thanks @mridulm @Ngone51

Reply:
> The cost for `getTaskAccumulableInfosAndProgressRate` should be similar to what we currently have already, right ?

Sorry, I commented in the wrong place. I actually meant the invocation of `setTaskRecordsAndRunTime()` in `handleSuccessfulTask()`.

Reply:
@mridulm your idea sounds good to me!